File: android/core/src/main/kotlin/com/negfeed/epyc/core/TurnEngine.kt
package com.negfeed.epyc.core

/**
 * Pure port of `src/providers/game-model/game-model.ts` turn/thread logic.
 * Behavior is asserted against `tools/golden/turn_vectors.json` (the same vectors
 * the iOS EpycCore tests use).
 */
object TurnEngine {

    /** One thread per player; `playerCount + 1` atoms alternating DRAWING/GUESS, all NOT_STARTED. */
    fun buildEmptyThread(playerCount: Int, word: String): GameThread {
        val atoms = (0 until playerCount + 1).map { i ->
            GameAtom(
                type = if (i % 2 == 0) GameAtomType.DRAWING else GameAtomType.GUESS,
                state = GameAtomState.NOT_STARTED,
            )
        }
        return GameThread(word = word, gameAtoms = atoms)
    }

    fun buildEmptyThreads(playerCount: Int, words: List<String>): List<GameThread> =
        (0 until playerCount).map { buildEmptyThread(playerCount, words[it % words.size]) }

    /** `(threadIndex - atomIndex + n) % n` (kept non-negative). */
    fun atomPlayerIndex(address: AtomAddress, playersCount: Int): Int {
        val n = playersCount
        return (((address.threadIndex - address.atomIndex) % n) + n) % n
    }

    /** For `i` in `0..playersCount`: `{ thread: (i + playerIndex) % n, atom: i }`. */
    fun playerAtomAddresses(playerIndex: Int, playersCount: Int): List<AtomAddress> {
        val n = playersCount
        val result = ArrayList<AtomAddress>(n + 1)
        var i = 0
        while (i < n + 1) {
            result.add(AtomAddress(threadIndex = (i + playerIndex) % n, atomIndex = i))
            i++
        }
        return result
    }

    fun getNextAtom(threads: List<GameThread>, usersOrder: List<String>, userId: String): NextAtom {
        val playerIndex = usersOrder.indexOf(userId)
        return getNextAtom(threads, usersOrder.size, playerIndex)
    }

    fun getNextAtom(threads: List<GameThread>, playersCount: Int, playerIndex: Int): NextAtom {
        for (address in playerAtomAddresses(playerIndex, playersCount)) {
            val thread = threads[address.threadIndex]
            val atom = thread.gameAtoms[address.atomIndex]
            val prev = if (address.atomIndex > 0) thread.gameAtoms[address.atomIndex - 1] else null
            if (atom.state != GameAtomState.DONE) {
                val ready = prev == null || prev.state == GameAtomState.DONE
                return NextAtom(address = address, readyToPlay = ready, allAtomsDone = false)
            }
        }
        return NextAtom(address = null, readyToPlay = false, allAtomsDone = true)
    }

    /** True when the last atom of every thread is DONE. */
    fun isGameDone(threads: List<GameThread>): Boolean =
        threads.all { it.gameAtoms.lastOrNull()?.state == GameAtomState.DONE }
}

File: ios/EpycCore/Sources/EpycCore/TurnEngine.swift
import Foundation

/// Pure port of `src/providers/game-model/game-model.ts` turn/thread logic.
/// Behavior is asserted against `tools/golden/turn_vectors.json`.
public enum TurnEngine {

    // MARK: Thread construction

    /// One thread per player; each thread has `playerCount + 1` atoms alternating
    /// DRAWING (even index) / GUESS (odd index), all NOT_STARTED.
    public static func buildEmptyThread(playerCount: Int, word: String) -> GameThread {
        var atoms: [GameAtom] = []
        for i in 0..<(playerCount + 1) {
            atoms.append(GameAtom(type: i % 2 == 0 ? .drawing : .guess, state: .notStarted))
        }
        return GameThread(word: word, gameAtoms: atoms)
    }

    /// `playerCount` empty threads. `words` must supply one word per thread.
    public static func buildEmptyThreads(playerCount: Int, words: [String]) -> [GameThread] {
        (0..<playerCount).map { buildEmptyThread(playerCount: playerCount, word: words[$0 % words.count]) }
    }

    // MARK: Rotation

    /// `(threadIndex - atomIndex + n) % n`
    public static func atomPlayerIndex(_ address: AtomAddress, playersCount n: Int) -> Int {
        ((address.threadIndex - address.atomIndex) % n + n) % n
    }

    /// For `i` in `0...playersCount`: `{ thread: (i + playerIndex) % n, atom: i }`.
    public static func playerAtomAddresses(playerIndex: Int, playersCount n: Int) -> [AtomAddress] {
        var result: [AtomAddress] = []
        var i = 0
        while i < n + 1 {
            result.append(AtomAddress(threadIndex: (i + playerIndex) % n, atomIndex: i))
            i += 1
        }
        return result
    }

    // MARK: Next atom

    /// First non-DONE atom along the player's address sequence. `readyToPlay` iff
    /// the previous atom in that thread is DONE; `allAtomsDone` when exhausted.
    public static func getNextAtom(threads: [GameThread], usersOrder: [String], userId: String) -> NextAtom {
        let playerIndex = usersOrder.firstIndex(of: userId) ?? -1
        return getNextAtom(threads: threads, playersCount: usersOrder.count, playerIndex: playerIndex)
    }

    public static func getNextAtom(threads: [GameThread], playersCount n: Int, playerIndex: Int) -> NextAtom {
        let addresses = playerAtomAddresses(playerIndex: playerIndex, playersCount: n)
        for address in addresses {
            let thread = threads[address.threadIndex]
            let atom = thread.gameAtoms[address.atomIndex]
            let prev: GameAtom? = address.atomIndex > 0 ? thread.gameAtoms[address.atomIndex - 1] : nil
            if atom.state != .done {
                let ready = (prev == nil) || (prev?.state == .done)
                return NextAtom(address: address, readyToPlay: ready, allAtomsDone: false)
            }
        }
        return NextAtom(address: nil, readyToPlay: false, allAtomsDone: true)
    }

    // MARK: Completion

    /// True when the last atom of every thread is DONE.
    public static func isGameDone(threads: [GameThread]) -> Bool {
        for thread in threads {
            guard let last = thread.gameAtoms.last, last.state == .done else { return false }
        }
        return true
    }

    /// Deterministic Fisher-Yates-style shuffle used on game start. The legacy code
    /// shuffles with `Math.random()`; production uses `Int.random`. A seeded variant
    /// is provided for tests/reproducibility.
    public static func shuffled(_ users: [String], using generator: inout some RandomNumberGenerator) -> [String] {
        var u = users
        u.shuffle(using: &generator)
        return u
    }
}
